pkg/distributor/level_detection.go

```diff
@@ -0,0 +1,197 @@
+	for _, field := range allowedFields {
+		allowedFieldsMap[field] = struct{}{}
+	}
+	return allowedFieldsMap
+}
+
+type LevelDetector struct {
+	validationContext validationContext
+}
+
+func (l *LevelDetector) shouldDiscoverLogLevels() bool {
+	return l.validationContext.allowStructuredMetadata && l.validationContext.discoverLogLevels
+}
+
+func (l *LevelDetector) extractLogLevel(labels labels.Labels, structuredMetadata labels.Labels, entry logproto.Entry) (logproto.LabelAdapter, bool) {
+	levelFromLabel, hasLevelLabel := l.hasAnyLevelLabels(labels)
+	var logLevel string
+	if hasLevelLabel {
+		logLevel = levelFromLabel
+	} else if levelFromMetadata, ok := l.hasAnyLevelLabels(structuredMetadata); ok {
+		logLevel = levelFromMetadata
+	} else {
+		logLevel = l.detectLogLevelFromLogEntry(entry, structuredMetadata)
+	}
+
+	if logLevel == "" {
+		return logproto.LabelAdapter{}, false
+	}
+	return logproto.LabelAdapter{
+		Name:  constants.LevelLabel,
+		Value: logLevel,
+	}, true
+}
+
+func (l *LevelDetector) hasAnyLevelLabels(labels labels.Labels) (string, bool) {
+	for lbl := range l.allowedLabelsForLevel() {
+		if labels.Has(lbl) {
+			return labels.Get(lbl), true
+		}
+	}
+	return "", false
+}
+
+func (l *LevelDetector) detectLogLevelFromLogEntry(entry logproto.Entry, structuredMetadata labels.Labels) string {
+	// otlp logs have a severity number, using which we are defining the log levels.
+	// Significance of severity number is explained in otel docs here https://opentelemetry.io/docs/specs/otel/logs/data-model/#field-severitynumber
+	if otlpSeverityNumberTxt := structuredMetadata.Get(push.OTLPSeverityNumber); otlpSeverityNumberTxt != "" {
+		otlpSeverityNumber, err := strconv.Atoi(otlpSeverityNumberTxt)
+		if err != nil {
+			return constants.LogLevelInfo
+		}
+		if otlpSeverityNumber == int(plog.SeverityNumberUnspecified) {
+			return constants.LogLevelUnknown
+		} else if otlpSeverityNumber <= int(plog.SeverityNumberTrace4) {
+			return constants.LogLevelTrace
+		} else if otlpSeverityNumber <= int(plog.SeverityNumberDebug4) {
+			return constants.LogLevelDebug
+		} else if otlpSeverityNumber <= int(plog.SeverityNumberInfo4) {
+			return constants.LogLevelInfo
+		} else if otlpSeverityNumber <= int(plog.SeverityNumberWarn4) {
+			return constants.LogLevelWarn
+		} else if otlpSeverityNumber <= int(plog.SeverityNumberError4) {
+			return constants.LogLevelError
+		} else if otlpSeverityNumber <= int(plog.SeverityNumberFatal4) {
+			return constants.LogLevelFatal
+		}
+		return constants.LogLevelUnknown
+	}
+
+	return l.extractLogLevelFromLogLine(entry.Line)
+}
+
+func (l *LevelDetector) extractLogLevelFromLogLine(log string) string {
+	logSlice := unsafe.Slice(unsafe.StringData(log), len(log))
+	var v []byte
+	if isJSON(log) {
+		v = l.getValueUsingJSONParser(logSlice)
+	} else {
+		v = l.getValueUsingLogfmtParser(logSlice)
+	}
+
+	switch {
+	case bytes.EqualFold(v, []byte("trace")), bytes.EqualFold(v, []byte("trc")):
```

Comment:
nit: make the byte slices a package variable to allocate them only once

Reply:
fixed 👍